src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,95 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double SCALE_FACTOR = 1.5;
+    private int size;
+    private Object[] elements;
 
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
     }
 
     @Override
-    public void add(T value, int index) {
-
+    public void add(T value) {
+        grow();
+        elements[size] = value;
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
-
+    public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index: " + index
+                    + " out of size " + size);
+        }
+        grow();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T value = (T) elements[index];
+        size--;
+        System.arraycopy(elements, index + 1, elements, index, size - index);
+        return value;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == elements[i] || element != null && element.equals(elements[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("this element does not exist " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
+    }
+
+    @Override
+    public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("index is not valid");
+        }
+    }
+
+    private void grow() {
```

Comment:
```suggestion
    private void growIfNeeded() {
```
The name should reflect inner logic. We do not grow every time here.